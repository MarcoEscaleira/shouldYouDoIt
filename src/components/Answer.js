import React from 'react'
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {actions} from "../store/decisions";


const Answer = ({handleState, data: {msg, img}}) => (
  <div className="answer">
    <h1 className="answer__title">{ (msg === 'loading') ? "Loading" : msg }</h1>
    <img 
      src={img}
      alt="Should You Do It gif"
      className="answer__image"
    />
    <button 
      onClick={handleState}
      className="answer__btn"
    >
      Clean
    </button>
  </div>
);

Answer.propTypes = {
  handleState: PropTypes.func,
  data: PropTypes.object
};

const mapStateToProps = state => ({
  data: state.data
});

const mapDispatchToProps = {
  handleState: actions.resetState
};

export default connect(mapStateToProps, mapDispatchToProps)(Answer);