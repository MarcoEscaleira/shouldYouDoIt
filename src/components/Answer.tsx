import React from 'react';
import { connect } from 'react-redux';
import { StoreState } from 'store';
import loadingGif from '~/assets/gifs/loading.gif';
import { actions, selectors, DecisionsStateData } from '../store/decisions';
// TODO: Check import export of types not working
import { ResetData } from '../store/decisions/actions';

export interface Props {
  resetData: ResetData;
  data: DecisionsStateData
}

const Answer: React.FC<Props> = ({
  resetData,
  data: { msg, img }
}) => (
  <div className="answer">
    <h1 className="answer__title">{msg}</h1>
    <img src={img || loadingGif} alt="Should You Do It gif" className="answer__image" />
    <button onClick={resetData} className="answer__btn">
      Clean
    </button>
  </div>
);

Answer.defaultProps = {
  data: { msg: 'loading', img: '' }
}

const mapStateToProps = (state: StoreState) => ({
  data: selectors.getData(state)
});

const mapDispatchToProps = {
  resetData: actions.resetData
};

// @ts-ignore
export default connect(mapStateToProps, mapDispatchToProps)(Answer);
