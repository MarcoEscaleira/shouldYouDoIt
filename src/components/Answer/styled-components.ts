// @ts-nocheck
import styled from 'styled-components';
import { sizes, pallete } from '../../styles/_settings';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  margin: ${sizes.s} 0 1rem 0;
  max-width: 30rem;
`;

export const Title = styled.h1`
  color: ${pallete.lightest};
  font-size: 2rem;
  border-left: 0.3rem solid darken(${pallete.lightest}, 20%);
  border-right: 0.3rem solid darken(${pallete.lightest}, 20%);
  padding-left: ${sizes.m};
  text-align: center;
  text-transform: capitalize;
`;

export const Image = styled.img`
  width: 30rem;
  height: 25rem;
  margin-bottom: ${sizes.m};
`;

export const CleanButton = styled.button`
  flex-grow: 1;
  background-color: ${pallete.third_dark};
  height: ${sizes.xl};
  color: ${pallete.white};
  border: none;
  border-radius: 0.5rem;
`;