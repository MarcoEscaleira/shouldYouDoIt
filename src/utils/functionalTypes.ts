export type NullaryFn<R> = () => R;
export type UnaryFn<A, R> = (a: A) => R;
export type BinaryFn<A, B, R> = (a: A, b: B) => R;
export type TernaryFn<A, B, C, R> = (a: A, b: B, c: C) => R;
export type QuadFn<A, B, C, D, R> = (a: A, b: B, c: C, d: D) => R;
export type QuinFn<A, B, C, D, F, R> = (a: A, b: B, c: C, d: D, f: F) => R;
export type SixFn<A, B, C, D, F, H, R> = (
  a: A,
  b: B,
  c: C,
  d: D,
  f: F,
  h: H
) => R;
