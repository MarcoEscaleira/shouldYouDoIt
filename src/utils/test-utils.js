import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import store from '../store';

export const renderWithRedux = (ui, initialState) => {
  return {
    ...render(
      <Provider store={store(initialState)}>
        <Router>{ui}</Router>
      </Provider>
    )
  };
};

export const renderWithRouter = ui => {
  return {
    ...render(<Router>{ui}</Router>)
  };
};

const middlewares = [thunk];
export const mockStore = configureStore(middlewares);
