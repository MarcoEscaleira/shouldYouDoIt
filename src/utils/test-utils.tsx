import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import store, { initialStoreState } from "../store";

export const renderWithRedux = (ui: JSX.Element, initialState?: {}) => {
  const state = {
    ...initialStoreState,
    ...initialState
  };
  return {
    ...render(
      <Provider store={store(state)}>
        <Router>{ui}</Router>
      </Provider>
    )
  };
};

export const renderWithRouter = (ui: JSX.Element) => ({
  ...render(<Router>{ui}</Router>)
});

const middlewares = [thunk];
export const mockStore = configureStore(middlewares);
