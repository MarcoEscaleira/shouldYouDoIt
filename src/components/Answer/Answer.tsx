import React from "react";
import { connect } from "react-redux";
import { StoreState } from "store";
import loadingGif from "../../assets/gifs/loading.gif";
import { Container, Title, Image, CleanButton } from "./styled-components";
import { actions, selectors, DecisionsStateData } from "../../store/decisions";
// TODO: Check import export of types not working
import { ResetData } from "../../store/decisions/actions";

export interface Props {
  resetData: ResetData;
  data: DecisionsStateData;
}

const Answer: React.FC<Props> = ({ resetData, data: { msg, img } }) => (
  <Container>
    <Title>{msg}</Title>
    <Image src={img || loadingGif} alt="Should You Do It gif" />
    <CleanButton onClick={resetData}>Clean</CleanButton>
  </Container>
);

Answer.defaultProps = {
  data: { msg: "loading", img: "" }
};

const mapStateToProps = (state: StoreState) => ({
  data: selectors.getData(state)
});

const mapDispatchToProps = {
  resetData: actions.resetData
};

// @ts-ignore
export default connect(mapStateToProps, mapDispatchToProps)(Answer);
