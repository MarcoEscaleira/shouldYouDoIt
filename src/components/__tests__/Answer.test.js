import React from 'react';
import userEvent from '@testing-library/user-event';
import { renderWithRedux } from '../../utils/test-utils';
import Answer from '../Answer';

describe('Answer test suite', () => {
  const initialState = {
    print: true,
    data: {
      img: 'aGifImg.gif',
      msg: 'do it'
    }
  };

  it('Should render Answer correctly', () => {
    const { container } = renderWithRedux(<Answer />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should render Answer with data', () => {
    const { container } = renderWithRedux(<Answer />, initialState);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should clean data after clicking on Clean btn', () => {
    const { container, getByText } = renderWithRedux(<Answer />, initialState);

    expect(container.firstChild).toMatchSnapshot();

    userEvent.click(getByText('Clean'));

    expect(container.firstChild).toMatchSnapshot();
  });
});
