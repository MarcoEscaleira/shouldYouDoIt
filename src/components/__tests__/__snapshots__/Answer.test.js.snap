// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer test suite Should clean data after clicking on Clean btn 1`] = `
<div
  class="answer"
>
  <h1
    class="answer__title"
  >
    do it
  </h1>
  <img
    alt="Should You Do It gif"
    class="answer__image"
    src="aGifImg.gif"
  />
  <button
    class="answer__btn"
  >
    Clean
  </button>
</div>
`;

exports[`Answer test suite Should clean data after clicking on Clean btn 2`] = `
<div
  class="answer"
>
  <h1
    class="answer__title"
  >
    loading
  </h1>
  <img
    alt="Should You Do It gif"
    class="answer__image"
    src="./gifs/Blocks-1s-200px.gif"
  />
  <button
    class="answer__btn"
  >
    Clean
  </button>
</div>
`;

exports[`Answer test suite Should render Answer correctly 1`] = `
<div
  class="answer"
>
  <h1
    class="answer__title"
  >
    loading
  </h1>
  <img
    alt="Should You Do It gif"
    class="answer__image"
    src="./gifs/Blocks-1s-200px.gif"
  />
  <button
    class="answer__btn"
  >
    Clean
  </button>
</div>
`;

exports[`Answer test suite Should render Answer with data 1`] = `
<div
  class="answer"
>
  <h1
    class="answer__title"
  >
    do it
  </h1>
  <img
    alt="Should You Do It gif"
    class="answer__image"
    src="aGifImg.gif"
  />
  <button
    class="answer__btn"
  >
    Clean
  </button>
</div>
`;
